parallel broker merges on fork join pool

clintropolis opened the pull request:
### Description

Implementation of proposal #8577 (see for more details).

In deviation from the proposal, in this PR this is now _on by default_. All query types except for `scan` are merged on a dedicated 'async' `ForkJoinPool`.

![ParallelMergeCombiningSequence](https://user-images.githubusercontent.com/1577461/66102488-64614800-e567-11e9-9fc3-f1dcf9473940.png)

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/incubator-druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/incubator-druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `ParallelMergeCombiningSequence`
 * `CachingClusteredClient`

clintropolis commented on processing/src/main/java/org/apache/druid/guice/LifecycleForkJoinPool.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+
+public class LifecycleForkJoinPool extends ForkJoinPool
```
This seems to be the cause of mysterious java 9+ compilation issues causing CI failures (my favorite part is how the compiler error doesn't at all specify what is causing the error). `ForkJoinPool` has the `@Contended` annotation, which from javadocs looks like it leaks to subclasses I guess?
```
 * <p>The class level {@code @Contended} annotation is not inherited and has
 * no effect on the fields declared in any sub-classes. The effects of all
 * {@code @Contended} annotations, however, remain in force for all
 * subclass instances, providing isolation of all the defined contention
 * groups. Contention group tags are not inherited, and the same tag used
 * in a superclass and subclass, represent distinct contention groups.
```

This might be fixed by something like:

```
<compilerArgs>
  <arg>--add-exports</arg>
  <arg>java.base/jdk.internal.misc=ALL-UNNAMED</arg>
</compilerArgs>
```

added to jdk9+ profile, but would also require removing all source/target directives into a jdk8 profile I think, since:

```
 error: option --add-exports not allowed with target 8
```
happens otherwise.

Not sure what is the best thing to do here. it could probably be worked around by I guess having a lifecycle'd module that wraps a `ForkJoinPool` for the provider instead of extending `ForkJoinPool`, since just using `ForkJoinPool` doesn't seem to cause this error in jdk9+, but I'm not sure if that is the better solution, or if we should try to add the compiler args.

leventov commented:
I withdraw from reviewing this, just one note: I'm strongly in favor of making the new behavior *opt-out* rather than opt-in, and removing the old behavior after a few versions of Druid released. We don't need more configurations here, parallel merges should be good for everybody.

The transition period of a few Druid versions is just a safety check against regressions in this feature which might be surfaced only in the production environment.

clintropolis commented:
>I'm strongly in favor of making the new behavior opt-out rather than opt-in, and removing the old behavior after a few versions of Druid released. We don't need more configurations here, parallel merges should be good for everybody.

@leventov I totally agree that this shouldn't have much of any config. Most of the config that exists in the PR currently is for my testing. I _was_ planning to make it opt-in initially though because I wasn't sure I would get sufficient testing in to be confident enough to make it the default yet, but I went ahead made the change in this branch and have been running all queries (except scan) on a test cluster for a couple days now and so far so good.

I'm a bit worried about the dumb, greedy, level of parallelism chooser currently in place, I need to perform additional stress testing to try and find where things fall apart. I think my goal here is to not be significantly _worse_ than the current behavior under load, though measuring this sounds .. difficult, so I will see what I can do.

jon-wei commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1059 @@
+      // adjust max to be no more than total pool parallelism less the number of running threads + submitted tasks
+      final int utilizationEstimate = getPool().getRunningThreadCount() + getPool().getQueuedSubmissionCount();
+      // minimum of 'max computed parallelism' and pool parallelism less current 'utilization estimate'
+      final int computedParallelism = Math.min(maxParallelism, getPool().getParallelism() - utilizationEstimate);
```
Is it possible to return a negative number from `getPool().getParallelism() - utilizationEstimate`? If so, it might be nicer, since that number is logged, to force it to 1 minimum to represent the serial processing case.

clintropolis replied:
It is, changed to be at least `1` for friendlier debug logs.

jon-wei commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1059 @@
+    private final T initialValue;
+    private final int yieldAfter;
+    private final int batchSize;
+    private final int depth;
```
Suggest adding a comment about how `depth` is the recursion count and is used for cumulative moving average smoothing as well

jon-wei commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1059 @@
+        } else if (cancellationGizmo.isCancelled()) {
+          // if we got the cancellation signal, go ahead and write terminal value into output queue to help gracefully
+          // allow downstream stuff to stop
+          outputQueue.offer(new ResultBatch<>());
```
Is it necessary to create a new object for the terminal value, or would it be possible to reuse a static instance?

clintropolis commented on processing/src/main/java/org/apache/druid/query/QueryToolChest.java:
```diff
@@ -113,9 +113,10 @@ public ObjectMapper decorateObjectMapper(final ObjectMapper objectMapper, final
    * function is used in the default {@link ResultMergeQueryRunner} provided by
    * {@link QueryToolChest#mergeResults(QueryRunner)} and can be used in additional future merge implementations
    */
+  @Nullable
```
Updated javadoc with new usage in this PR, describe implications of returning `null`, also added precondition checks for `ResultMergeQueryRunner` comparator and mergeFn factories.

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1059 @@
+    this.orderingFn = orderingFn;
+    this.combineFn = combineFn;
+    this.hasTimeout = hasTimeout;
+    this.timeoutAt = System.currentTimeMillis() + timeoutMillis;
```
Not :100:, but went ahead and made it consistently use `System.nanoTime`

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1059 @@
+        // if we have a small number of sequences to merge, or computed paralellism is too low, do not run in parallel,
+        // just serially perform the merge-combine with a single task
+        if (sequences.size() < 4 || parallelTaskCount < 2) {
+          LOG.info(
```
I definitely agree, this was at info level for my testing purposes, moving back to debug, as well as added back some additional debug level logs. I think in the future I would like to add some metrics for some of this stuff to help aid in tuning parameters, but maybe in a follow-up PR.

jihoonson commented on server/src/main/java/org/apache/druid/guice/DruidProcessingModule.java:
```diff
@@ -135,6 +136,26 @@ public ExecutorService getProcessingExecutorService(
+  {
+    return new LifecycleForkJoinPoolProvider(
+        config.getNumThreadsMergePool(),
+        ForkJoinPool.defaultForkJoinWorkerThreadFactory, // todo: ?
```
I guess this comment is related to a custom prioritized factory? Please update the comment.

clintropolis replied:
Ah, no this was for whether or not I was going to set custom thread names as @drcrallen did in #5913, see here https://github.com/apache/incubator-druid/pull/5913/files#diff-fa77f3240f159148c9eac7faa94c516fR47. I've decided against it for now and removed the todo since all of the logs are debug level other than the unhandled exception handler I just added, but am open to discussion.

jihoonson replied:
Maybe still it could be worth to be more debug friendly for who are not familiar with Druid internal.

jon-wei commented on docs/configuration/index.md:
```diff
@@ -1435,6 +1435,12 @@ The broker uses processing configs for nested groupBy queries. And, if you use g
 |`druid.processing.fifo`|If the processing queue should treat tasks of equal priority in a FIFO manner|`false`|
 |`druid.processing.tmpDir`|Path where temporary files created while processing a query should be stored. If specified, this configuration takes priority over the default `java.io.tmpdir` path.|path represented by `java.io.tmpdir`|
+|`druid.processing.useParallelMergePool`|Enable automatic parallel merging for Brokers on a dedicated async ForkJoinPool. If `false`, instead merges will be done serially on the `HTTP` thread pool.|`true`|
+|`druid.processing.numMergePoolThreads`|Size of ForkJoinPool|`druid.processing.numThreads * 1.5`|
```
Suggest mentioning that the default value is rounded up

clintropolis replied:
added

jihoonson commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+      final int computedParallelism = Math.min(maxParallelism, getPool().getParallelism() - utilizationEstimate);
+      // compute total number of layer 1 'parallel' tasks, the final merge task will take the remaining slot
+      final int computedOptimalParallelism = Math.min(
+          (int) Math.floor((double) sequences.size() / 2.0),
```
Would you add a comment where `2.0` came from?

clintropolis replied:
It came from needing at least 2 sequences per partition so there is something for the thread to merge.

himanshug commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+
+        // if we have a small number of sequences to merge, or computed paralellism is too low, do not run in parallel,
+        // just serially perform the merge-combine with a single task
+        if (sequences.size() < 4 || parallelTaskCount < 2) {
```
started skimming through the code here, so my understanding might be very superficial right now.

but, it seems sequences.size() check can be done in `CachingClusteredClient` and not use the parallel flow at all and let jetty thread do the work ... doing the check here means work is still done async for no reason ... no ?

clintropolis replied:
well, it is not for no reason, it is to have (nearly) all work being done on the same thread pool instead of having 2 competing thread pools of different sizes doing the work

clintropolis replied:
Ideally I would make scan queries run on this too, so that way all merges are done on the same pool, similar in spirit to the processing pools in the historicals i guess.

himanshug replied:
got it, makes sense.

clintropolis replied:
Also, thanks for having a look!

himanshug replied:
BTW, I think the check should be `sequences.size() <= 4` because following code

```
      final int computedOptimalParallelism = Math.min(
          (int) Math.floor((double) sequences.size() / 2.0),
          computedParallelism - 1
      );
```
would always return a value less/equal to 1 for sequences.size() == 3 .

clintropolis replied:
Ah, I think that does make the check against `sequences.size()` done after this redundant:

```
        final int parallelTaskCount = computeNumTasks();

        // if we have a small number of sequences to merge, or computed paralellism is too low, do not run in parallel,
        // just serially perform the merge-combine with a single task
        if (sequences.size() < 4 || parallelTaskCount < 2) {
...
```
and can just use parallelTaskCount.

himanshug commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+    protected void compute()
+    {
+      try {
+        final int parallelTaskCount = computeNumTasks();
```
what if multiple queries arrive concurrently, there would be races and they might overestimate available parallel capacity ?

clintropolis replied:
currently it can definitely overestimate how parallel it can go, though in my stress testing so far it doesn't seem to actually matter that much, as each fork join pool task runs for a short period so that the work is still all interleaved, just done slightly slower since there are more tasks to run.

If we wanted to try and only allocate as much parallelism as there are threads in the pool then we would need an external mechanism to share between all queries I think beyond asking the fork join pool directly, that grabs a number of slots from a fixed size counter and returns them when completed. Maybe something like the thread holders in @jihoonson previous attempt at parallel merges. This is because asking the fork join pool itself isn't adequate, due to the nature of how this operates where there is only 1 task per query at the time this is computed, but after that initial task is done it then executes all of the parallel tasks to do the actual merge work. Like imagine 3 queries come in, it launches 3 tasks to determine how parallel to go, and all ask the pool what the usage is and only see the 3 tasks, so they each think they can use all the remaining capacity of the pool.

Anyway, I was hoping to defer doing something like this until it becomes obvious that we actually need it, which I'm not certain that we do yet.

clintropolis replied:
All that said, i certainly there are smarter ways we could compute the level of parallelism and also how the input sequences are partitioned, the current approach is very greedy.

himanshug replied:
thanks , glad to know that it has been given considerable thought.

can we emit a metric for number of sequences handled per partition i.e. `sequences.size()/partitions` , as this code runs, It will be nice to be able to have visibility into this.

clintropolis replied:
:100: on board with adding metrics, I really want them to help me tune it, have just been on the fence if I should do it in this PR or a follow-up.

I'm still thinking about what all I would like to see, definitely partition sizes, number of partitions, number of total tasks required to fully process the result sets (recursion depths), maybe more, though I need to weigh this probably with what is useful to help me tune stuff now with what is actually useful long term, which would probably be translating these things I want to see into something an operator could use to tell how saturated their brokers are to aid in scaling decisions.

himanshug replied:
yes, my proposed metric was to see, for me as user, basically how much parallelism is really happening. actually I would add one more, periodically, number of queries received and number of queries parallelized.

I am OK if you wanna do that in a later PR , it should just be done before this code is shipped in a release.
more metrics could of course be there for devs to learn about reality and change code based on that.

on the scaling (I'm assuming you mean increasing number of broker boxes), I think the decision should be made based on much higher level metrics such as cpu/memory utilization of the boxes etc , but anyways that is speculation at this point . that said, I did write a k8s operator which I am hoping to open source soon (currently it only does the deployment) :)

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+
+    @Nullable
+    private final Queue<E> values;
+    private final boolean isTerminal;
```
Sure, that would work I guess, can change.

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+              count++;
+              if (count % batchSize == 0) {
+                yield();
+              }
```
I can put the yield after the value is added to the batch, if it makes it look less dicey. I'm not sure what a comment would look like? This block seems obvious to me, but maybe I'm too far in and can't properly gauge such things.

jihoonson replied:
I guess maybe the comment can explain how `accumulated` is used and what happens when `yield()` is called. For example, it wasn't clear to me how `ResultBatch` is created because `in` is always added to the same `accumulated`. Regarding `yield()`, it wasn't clear what code will be executed when `yield()` is called.

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+      values.offer(in);
+    }
+
+    public E get()
```
Wouldn't it return `ResultBatch<E>` if it looked like it returned a whole batch? I called it `get` to be consistent with `BatchedResultsCursor.get` which is just a wrapper around `ResultBatch.get`. I can name `getResult` I guess if `get` is bothersome.

jihoonson replied:
👍

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+        return true;
+      }
+      if (resultBatch == null || resultBatch.isDrained()) {
+        final Yielder<ResultBatch<E>> nextYielder = yielder.next(resultBatch);
```
It seemed more efficient to re-use the same batch since it would result less garbage objects and less pressure on the garbage collector, and not particularly less readable to me at least. Can I flip the question, is there a good reason to make a new object when I have a perfectly good one already allocated and ready to go?

clintropolis replied:
After thinking more about this, I agree it is a bit more obvious to use a new batch. I'll run benchmarks to make sure there isn't any performance degradation and make the change.

jihoonson replied:
My main concern was about if it's violating any concurrent access. I think there's no concurrency issue but creating new batches will be easier to find/remove concurrency issues if they are added in the future. But if there was any performance issue, then it would make more sense to reuse the same instance.

jihoonson replied:
I'm also wondering if the performance can be better if `ResultBatch` is reused and annotated with `@Contended`. Apparently, contended objects are allocated more space and so they can be isolated without polluting cache lines.

This will need some experiments, so I don't think my comment should be blocker of this PR.

clintropolis replied:
There appears to be no performance issues with creating a new object, so I will change it to be more in spirit of how the yielder is working, where a new one is returned by next.

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+      return item == null;
+    }
+
+    void addItem(E item)
```
removed

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+            // else, push accumulated value to the queue, accumulate again with next value as initial
+            outputBatch.add(currentCombinedValue);
+            batchCounter++;
+            if (batchCounter >= batchSize) {
```
There is no `outputBatch.size()`, and `outputBatch.values` which does have a `size` is an `ArrayDequeue` whose size seemed mildly less efficient than just keeping my own counter I guess:

```
public int size() {
        return (tail - head) & (elements.length - 1);
    }
```

I can change it to use the array dequeue size if you prefer, or I guess push my own counter into `ResultBatch` so it has a size, but less method overhead will generally always offer a slight performance advantage.

jihoonson replied:
Ah my comment was about counting in `ResultBatch` and adding `size()` method to it. My comment is a nit thing and you can freely ignore it.

clintropolis commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1077 @@
+   * task completes and executes a new task to continue where it left off. This value is initially set by the
+   * {@link MergeCombinePartitioningAction} to a default value, but after that this process is timed to try and compute
+   * an 'optimal' number of rows to yield to achieve a task runtime of ~10ms, on the assumption that the time to process
+   * n results will be approximately the same. 'recursionDepth' is used to track how many times a task has continued
```
`yieldAfter` is also in javadoc but not as a link

himanshug commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1071 @@
+          );
+          final double nextYieldAfter = Math.max(10.0 * ((double) yieldAfter / elapsedMillis), 1.0);
+          final double cumulativeMovingAverage = (nextYieldAfter + (recursionDepth * yieldAfter)) / (recursionDepth + 1);
+          final int adjustedNextYieldAfter = (int) Math.ceil(cumulativeMovingAverage);
```
assuming we really want task to preempt itself in about 10ms, I wonder why we couldn't just check at the `elasped_time_so_far` inside the loop whenever one batch was pushed i.e. inside the `if (batchCounter >= batchSize) {` block at line-495 .  if  `elasped_time_so_far` was greater than `10 ms` then break from the loop at that point? that would probably more accurately evict the task in about `10 ms`

also, it would be nice to know your mental notes on why you chose threshold to be `10 ms` .

himanshug replied:
maybe it would've been too expensive to do on each batch completion ... but in that case it could be checked on every, say 10 batch completion.

not saying that this necessarily needs to change but trying to understand how the choice is made.

clintropolis replied:
The 10ms is from @leventov who came up with this approach in general in [a thread in a previous attempt at parallel broker merges](https://github.com/apache/incubator-druid/pull/6629#discussion_r241089247), and I rolled with it because it seemed like a reasonable time period to get a decent amount of work done without starving other queries from running their tasks on the pool. Disclaimer that I haven't done a lot of measuring with other task running time targets since this one has been benchmarking reasonably well.

I haven't measure the difference, but I don't think it needs to be incredibly strict, was just shooting for close enough within reason, so it seemed cheaper to check the actual elapsed time less frequently. But agree it might be worth investigating, if I find some time I'll try to play around with it.

himanshug replied:
Idea is pre-emption is to give way to other work submitted to the FJP, so it should take `fjp.getQueuedSubmissionCount()` into consideration. If `fjp.getQueuedSubmissionCount() = 0` , that means there is no work pending on cpu availability, so there is no point in giving up the cpu, it should continue going.

Now, for the other case when there is queued work waiting for CPU, then we definitely wanna pre-empt, now the choice of 10ms vs 100ms is arbitrary and based purely on gut feeling. I would've chosen it to be 100ms just to err on the side of making same combine operation more efficient and not hogging CPU for a ridiculously long time.

Now, when we talk about 10ms above .. we want to measure the amount of time combine op got actually being on the CPU. But, in the while loop I see, there are many places where we are making potentially blocking operations .. so 
```
          final long elapsedMillis = Math.max(
              TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS),
              1L
          );
```
is giving us potentially way more than the time spent on CPU and `ThreadMXBean.getCurrentThreadCpuTime()` could be used to get that more accurately instead of `System.nanoTime()` .  (I am assuming that call to `ForkJoinPool.managedBlock(..)` doesn't change thread but comparing currentThreadID at start and end, and printing a warning if threadID changed would be good to know if that assumption is incorrect).

does it make sense ?

clintropolis replied:
>Idea of pre-emption is to give way to other work submitted to the FJP, so it should take fjp.getQueuedSubmissionCount()

My worry about using this and other fjp inspection methods too frequently was that they look semi expensive, looping over the work queues to count stuff, but I can look into it a bit more and do some experiments, as maybe measurement overhead is insignificant.

>The choice of 10ms vs 100ms is arbitrary and based purely on gut feeling.

Yes this is true, I think I could likely make it configurable to assist in tuning.

>Now, when we talk about 10ms above .. we want to measure the amount of time combine op got actually being on the CPU. But, in the while loop I see, there are many places where we are making potentially blocking operations 

Yeah, I guess I was more approaching it as time slicing the thread pool to try to minimize the number of extra threads the pool needs to create for normal operation when stuff is blocking, rather than directly considering the cpu usage, but what you are suggesting might make more sense because that is measuring the actual work it is doing, I don't have a strong gut feeling on either way so did the easiest approach to measurement. 

At least the way stuff is currently structured, I don't believe I can currently fully take the potentially blocking parts out of the time measurement unless I do a lot more, smaller measurements and combine them, measuring only periods where I know blocking will not happen. 

I will do some more experiments and see what I can do to determine what is the most effective way to do the measurements, what actually should be measured, as well as how they should be used in terms of task run time (though easier said than done I think).

himanshug replied:
I know giving ideas is easy and  totally agree about the "easier said than done" part , and you have the first right to any choice or to degree of experimentations done as the author :)

Looked at `fjp.getQueuedSubmissionCount()` and indeed it is fairly expensive at busy times, OTOH noticed that there is a `fjp.hasQueuedSubmissions()` which is much faster and just tells if `fjp.getQueuedSubmissionCount() > 0` 

> At least the way stuff is currently structured, I don't believe I can currently fully take the potentially blocking parts out of the time measurement unless I do a lot more, smaller measurements and combine them, measuring only periods where I know blocking will not happen.

yeah, Even if you did that , cpu time is more effective because so called "blocking" code in `ManagedBlocker` isn't necessarily blocking from OS perspective every time and could consume cpu time.

leventov replied:
Regardless of what value is chosen (or a config parameter introduced), there should be an extensive comment describing the default choice. Currently, there is a magic constant.

[Make the code obvious by adding **a lot** of explaining comments](https://github.com/apache/incubator-druid/blob/fe5bb66aad5ba3ce808a7bf243a6ef24240d7507/dev/principles.md#explaining-comments).

clintropolis replied:
>I am assuming that call to ForkJoinPool.managedBlock(..) doesn't change thread but comparing currentThreadID at start and end, and printing a warning if threadID changed would be good to know if that assumption is incorrect

I dug around deeper into the code for `ForkJoinPool` and for good measure added the `log.warn` and did some tests, and thread id does appear stable through a managed block, which makes sense to me. From my understanding, work does not move to a new thread, it is just that the pool just no longer counts the thread as 'active', allowing the pool to spawn a new thread which presumably can actually do work while this one chills out. The blocked thread becomes active again, as far as the pool is concerned, once the managed blocker becomes release-able. Then the pool will try to shrink the overall thread count back down as it is able.

clintropolis replied:
Thanks for digging in on this @himanshug and making me really have a closer look at this area in particular, since I think it really is one of the most important parts to ensure that the pool operates well when it is heavily saturated. I've been working on a benchmark that directly tests `ParallelMergeCombiningSequence` in order to try out the differences between 10ms and 100ms, using wall time vs using cpu time, behavior when overloaded, etc. 

While I haven't quite got to that yet, in the process I have discovered a mistake that heavily impacted performance when the pool was overloaded when computing the level of parallelism: https://github.com/apache/incubator-druid/pull/8578/files#diff-84792f9d3cefe47cbb471669dce2a276R374, specifically that it was using the pool parallelism instead of the processor count, which meant it was overestimating it's capacity for work. These benchmarks are not pushed yet, and will probably change a lot as I clean them up, but the rough explanation is

* `testJustOne` uses `ParallelMergeCombiningSequence` and a single jmh thread on empty fork join pool
* `testManyParallel` does 24 jmh threads with `ParallelMergeCombiningSequence` all being evaluated at the same time (this was my suspected worst case for the greedy parallelism chooser)
* `testManyParallelJitter` does 24 jmh threads with a random 0-50ms delay before starting work on the `ParallelMergeCombiningSequence` to give a more realistic load pattern where not everything is starting at exactly the same time
* `testSerial` uses `CombiningSequence` wrapping a `MergeSequence` to do equivalent operation to `ParallelMergeCombiningSequence` 
* `testSerialParallel` is 24 threads of `testSerial`
* `testSerialParallelJitter` is `testSerialParallel` with 0-50ms of random jitter

Benchmarks were run twice both with and without synchronized benchmark iterations. Synchronized to show performance of the exact scenario described for each benchmark above, and without synchronization which lets each thread do their iterations at their leisure and more likely emulating an actual sustained heavy load pattern. 

```
pool parallelism - running + queued
(no sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt     Score     Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20    60.393 ±   4.754  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  1389.137 ± 904.866  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20   647.165 ± 168.406  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20    75.212 ±   0.600  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20   509.219 ±   9.345  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20   505.975 ±  12.986  ms/op

(sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt     Score      Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20    61.379 ±    5.957  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  1883.554 ± 1614.321  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20   540.248 ±   17.194  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20    76.956 ±    0.768  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20   515.362 ±   14.206  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20   507.267 ±   10.427  ms/op
```

after the change to `final int computedParallelism = maxParallelism - utilizationEstimate`:

```
num cores - (running + queue)
(no sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt    Score    Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20   49.720 ±  1.188  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  533.131 ± 14.251  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20  539.346 ± 13.348  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20   75.876 ±  0.579  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20  500.047 ± 12.856  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20  517.185 ± 11.089  ms/op

(sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt    Score    Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20   49.861 ±  1.088  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  522.590 ± 16.520  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20  530.802 ± 18.761  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20   75.877 ±  0.882  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20  517.718 ± 12.978  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20  520.770 ±  9.040  ms/op
```

After the modification, things look a lot better. I'm going to clean up the benchmarks and probably split out the same threaded combining sequence wrapping a merge sequence reference benchmarks so they don't get run with parameters that don't apply, which will allow me to dig deep into the tuning matters about target task run time and how to measure it, batch size, as well as some additional variations like sequences with occasional blocking periods and stuff.

himanshug commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -0,0 +1,1071 @@
+import java.util.function.BinaryOperator;
+
+/**
+ * Artisanal, locally-sourced, hand-crafted, gluten and GMO free, bespoke, small-batch parallel merge combinining sequence
```
not organic ? :-P

maybe copy some part from https://github.com/apache/incubator-druid/issues/8577 also and add link to that .

clintropolis replied:
heh, thanks for reminding me to add some actually useful text here, I was also going to mention that functionally it is like a composite of `MergeSequence` and `CombiningSequence` :+1:

clintropolis commented:
Ok, apologies for going dark while I was running benchmarks and making minor adjustments over the last couple weeks; I have some good findings and I'll try to capture everything here across a few comments to follow.

**tl;dr** though is that I have changed the defaults to `100ms` run time (using CPU time as the basis of measurement), batch sizes of `4096` results, and `16384` as the number of results to process (4 batches) for the initial task. Based on my observations, this configuration seems to perform the best across different result set sizes, blocking patterns, and concurrency loads.

I've also gained more confidence that this will be an acceptable default for most workloads where the broker is not expected to be constantly overloaded with significantly more active requests than the number of cores it has, though this approach can even be tuned to run reasonably well in this case. Next we can see if all of you will come to the same conclusion :+1:

clintropolis commented:
### ParallelMergeCombiningSequenceBenchmark and ParallelMergeCombiningSequenceJmhThreadsBenchmark
These are the new benchmarks I have added which I have been using to try and illustrate some things so that we can feel confident about merging this PR and making it the default behavior.

**My primary goal with this PR and approach to doing broker merges in parallel is to:** 
  * be better when we can (low concurrency/under utilized processors)
  * be not much worse when we can't (high concurrency/over utilized processors)
  
so I have been mainly focusing on the performance increase to be had in the best case, and on trying to trace out the behavior in the worst case so we can see that it is not _dramatically_ worse than what is currently in master.

For each of these sets of observations, I decided to try different configurations in order to determine the best set of defaults based on the earlier questions about whether 10ms or 100ms task run time would be better, wall-time or CPU time, etc. 

The `strategy` parameter of the benchmark encodes these configuration options, with parallel merges in the form of 

  `parallelism-{number of parallel tasks}-{target task time}ms-{batch size}-{initial task yield size}`

and `combiningMergeSequence-same-thread` to be the reference that is the same threaded analog of what `ParallelMergeCombiningSequence` is doing. Additionally, `inputSequenceType` specifies the behavior of the input sequences being fed into the `strategy`, covering behavior when the sequences are totally non-blocking, non-blocking after some amount of delay, and occasionally blocking after some delay. 

The benchmarks are split into two because I wanted to look at two slightly different measurements of the worst case. The first is provided by `ParallelMergeCombiningSequenceBenchmark` which is measuring the time to complete a concurrently executed group of queries which uses an executor thread pool to run some number of concurrent sequences and collect the average time for the entire group of threads to complete. The second is to focus more on invididual thread experience and uses the JMH `@Thread` annotation, and is provided by `ParallelMergeCombiningSequenceJmhThreadsBenchmark`. This measurement is expected to be lower than the concurrency performance measurement provided by `ParallelMergeCombiningSequenceBenchmark` because it is not driven entirely by the _slowest_ thread, rather the slower threads will be averaged across all of the threads.

If you would like to run these yourself:
```
java -server -cp benchmarks/target/benchmarks.jar org.apache.druid.benchmark.sequences.ParallelMergeCombiningSequenceBenchmark
```
and
```
java -server -cp benchmarks/target/benchmarks.jar org.apache.druid.benchmark.sequences.ParallelMergeCombiningSequenceJmhThreadsBenchmark
```

Just be sure to adjust the configuration to **not include all of the available parameters** unless you have multiple days of patience for them to complete.

clintropolis commented:
### Hyper-threading and default ForkJoinPool parallelism and per query parallelism configuration values

Another side-note before I dig into benchmark results, I wanted to call out that I modified the default values for `druid.processing.merge.pool.parallelism` and `druid.processing.merge.pool.defaultMaxQueryParallelism` to be 0.75 and 0.5 times `Runtime.getRuntime().availableProcessors()` instead of 1.5 and 1.0 respectively, to make an assumption that there are 2 hyper-threads per physical core.

This run of `ParallelMergeCombiningSequenceBenchmark` on an `m5.4xl`, which has 8 physical cores but reports 16 cores through `Runtime.getRuntime.availableProcessors()` since that is the 'vCPU' count, illustrates the reason for this.

```
Benchmark                                     (concurrentSequenceConsumers)  (inputSequenceType)  (numSequences)  (rowsPerSequence)                          (strategy)  Mode  Cnt     Score    Error  Units
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000  combiningMergeSequence-same-thread  avgt   20  2975.944 ±  3.406  ms/op
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000         parallelism-1-10ms-256-1024  avgt   20  2903.598 ±  7.425  ms/op
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000         parallelism-4-10ms-256-1024  avgt   20   918.689 ±  7.172  ms/op
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000         parallelism-6-10ms-256-1024  avgt   20   549.228 ±  1.545  ms/op
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000         parallelism-8-10ms-256-1024  avgt   20   394.795 ±  1.110  ms/op
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000        parallelism-12-10ms-256-1024  avgt   20   707.393 ±  2.392  ms/op
ParallelMergeCombiningSequenceBenchmark.exec                              1         non-blocking             128              75000        parallelism-16-10ms-256-1024  avgt   20  2069.897 ± 35.939  ms/op
```

This is of course because hyper-threads are not _real_ cores, so when actually fully utilizing a CPU to do work, counting them as such results in slower overall throughput. This is in line with the fine print on hyper-threads I think. 

This assumption for the default value might be wrong, but it held true for both my laptop and AWS, so maybe isn't unreasonable. I updated the user facing documentation on these settings to include this assumption, so that if an operators hardware differs and has more or no hyper-threading the value should be adjusted accordingly.

clintropolis commented:
### best case
All benchmarks for this phase were done on an AWS `m5.8xl` which has 16 physical cores, 32 hyper-threads. The currently chosen default configuration is `parallelism-16-100ms-4096-16384` in all of the following charts.

As mentioned in an earlier comment, the first area I wanted to benchmark is looking at the best case scenario to illustrate the raw performance increase of performing sequence and result merging in parallel when no other work is concurrently happening. 

Relatively quickly the benchmarks made it apparent that the parallelism chooser (`computeNumTasks`) was far too greedy, which in hind-sight seems obvious. Prior to running this benchmark, `computeNumTasks` did a check to ensure that _at least 2 sequences would be in every partition_ since that is the minimum number required to do meaningful work in parallel. However, this leads to a potentially very unbalanced amount of work between the layer 1 and layer 2 tasks, where layer 1 tasks would do very little work but the final merge layer 2 task would have a lot more to do from the increased number of input sequences. See this chart, which is plotting out artificially limited parallelism to compare to 'max' parallelism with different configurations, how the more parallelism used results in slower overall times.

![parallelism-m5 8xl-original](https://user-images.githubusercontent.com/1577461/67904227-f9d20680-fb2a-11e9-8d52-c3bfcb7f39a3.gif)

By moving to balance the number of partitions to the size of each partition by limiting parallelism to the square root of total number of input sequences, this has been improved to the point that I consider 'good enough' for now.

![parallelism-m5 8xl-variants](https://user-images.githubusercontent.com/1577461/67904252-06565f00-fb2b-11e9-8104-16202f7cd78f.gif)


These performance improvements should be most apparent in
* large clusters (high number of sequences to merge) with most result set sizes
* moderate sized clusters with large result sets and high combinability

This performance improvement applies across all result set sizes as well (here from 1k per input sequence up to 10m per input sequence)

![parallelism-m5 8xl-final](https://user-images.githubusercontent.com/1577461/67904268-0eae9a00-fb2b-11e9-81cc-5fdec8554d75.gif)


Of note, in the best case the configuration parameters other than maximum parallelism seemed to provide a relatively insignificant amount of variation in these tests, leading me to believe that they didn't really matter so much. For small result sets, the default value I have chosen does tend to do worse than the smaller sized tasks, but performs better with larger result sets which I believe is what we should tune for. 

However, the worst case, heavy concurrency benchmarks did end up changing my stance on this, where the larger sized unit of work per task seems to perform a lot better, particularly when blocking is involved.

clintropolis commented:
### worst case
The other really important part to highlight, is to determine how this approach degrades under increasing concurrency by looking at the absolute worst case scenario with the current approach, the concurrent load spike. I consider this the nemesis of the current parallelism chooser, since it will potentially dramatically under-estimate the available CPU capacity. 

For reference, I included a measurement when the `ParallelMergeCombiningSequence` level of parallelism was limited to 1. This is what the computed level of parallelism will trend towards when load is gradually increased and sustained, and so is how continued overloading is expected to behave.

From the left of the following line plots, the vertical dotted lines represent the number of physical cores and number of hyper-threads of the benchmark machine.

![concurrency-m5 8xl-burst](https://user-images.githubusercontent.com/1577461/67905082-8c73a500-fb2d-11e9-8f7c-baecc07fc8e9.gif)

This looks pretty bad for parallelism under huge load spikes, but also keep in mind that this measurement is looking at the time to complete all threads, not the average per thread timing (so at the far right of the plot, the time to simultaneously complete 4 times as many threads as we have processors).

Using an offset of 10ms before each concurrent thread to help even out the load shrinks the gap of performance degradation significantly:
![concurrency-m5 8xl-10ms-offset-burst](https://user-images.githubusercontent.com/1577461/67904913-0e170300-fb2d-11e9-9e0b-24c8c076aff8.gif)
and helps tie the greedy parallelism chooser to the poor load spike performance.

#### blocking
Looking at the `ParallelMergeCombiningSequenceJmhThreadsBenchmark` output shows similar observations, but the apparent degradation doesn't seem as bad because it is not entirely driven by the slowest thread.

![concurrency-m5 8xl-non-blocking-threaded](https://user-images.githubusercontent.com/1577461/67905376-5c78d180-fb2e-11e9-9e06-153180c8d738.gif)
 

This is due to many of the concurrent threads not seeing particularly worse performance, but having the worst performing threads averaged into the overall time. It also produces a plot that grows in the same linear fashion as the same threaded approach currently in use.



Introducing blocking input sequences caused some unexpected erratic behavior in the 10ms target task run time that I found surprising, but was repeatable. As far as I have been able to determine, the increased overall number of tasks that need to run on the `ForkJoinPool` and additional contention since there are more operations adding to and taking from a `BlockingQueue`.
![concurrency-m5 8xl-initially-blocking-100-500ms](https://user-images.githubusercontent.com/1577461/67904793-afea2000-fb2c-11e9-821a-4a7a2a23bdd1.gif)
The larger task run time results in the least amount of both tasks and cooperative blocking operations, which is why it was ultimately chosen as the default.

Longer delays and smaller result sets also shrink the apparent difference between the approach in this PR and the current approach, even without delays between threads:
![concurrency-m5 8xl-initially-blocking-4000-5000ms](https://user-images.githubusercontent.com/1577461/67905953-5a177700-fb30-11e9-8e87-e7f1999161d9.gif)

clintropolis commented:
### my conclusions
I think the performance degradation under load spikes falls within what I consider acceptable range.  Particularly important, I find the degradation to be relatively predictable so that we can update the cluster tuning guide (in a future PR) to best advise how to tune your brokers for a given cluster size and load pattern. 

I also believe there is quite a lot of room for further improvement, so maybe we could get to place where this approach nears the same overloaded characteristics of the http thread pool based merges we currently perform.

jihoonson commented:
Thank you for the detailed benchmark results! It looks great but I wonder how the default configuration works under more realistic load. For example, it would be more realistic if there are like 80% of light queries and 20% of heavy queries that have a shorter delay and a larger delay, respectively.

> I think the performance degradation under load spikes falls within what I consider acceptable range. Particularly important, I find the degradation to be relatively predictable so that we can update the cluster tuning guide (in a future PR) to best advise how to tune your brokers for a given cluster size and load pattern.

Sounds nice.

clintropolis commented:
>Thank you for the detailed benchmark results! It looks great but I wonder how the default configuration works under more realistic load. For example, it would be more realistic if there are like 80% of light queries and 20% of heavy queries that have a shorter delay and a larger delay, respectively.

This sounds good. I think i went a bit hard on this PR in the benchmarks I have presented so far in terms of targeting the worst cases which aren't super realistic, which I think maybe looks a lot scarier than a typical heavy load will appear practice. The existing worst case benchmarks are basically depicting what happens if a bunch of moderate to large result set sized queries _all happen simultaneously_ and even more _all simultaneously have work to do instead of some of them blocking waiting for input_, 

I'l throw together another benchmark to try and plot out a more realistic heavy load case to see how that looks.

clintropolis commented:
### more realistic worst case
I reworked the JMH thread based benchmark to use thread groups to examine what happens in a more realistic scenario, with the newly renamed `ParallelMergeCombiningSequenceThreadedBenchmark`. I find this benchmark to be a fair bit less scary than the previous 'worst case' benchmarks, which focused on an impossible scenario because I really wanted to dig in and see where and how the wheels fell off.

This benchmark models a more 'typical' heavy load, where the majority of the queries are smaller result-sets with shorter blocking times and a smaller subset are larger result sets with longer initial blocking times. By using thread groups we can look at performance for these 'classes' of queries as load increases. 

This set was collected with a ratio of 1 'moderately large' query for every 8 'small' queries, where 'moderately large' is defined as input sequence row counts of 50k-75k rows and blocking for 1-2.5 seconds before yielding results, and 'small' is defined as input sequence row counts of 500-10k and blocking for 50-200ms. Keep in mind while reviewing the result that I collected data on a significantly higher level of parallelism than I would expect a 16 core machine to be realistically configured to handle. I would probably configure an m5.8xl with ~64 http threads, but collected data points up to 128 concurrent sequences being processed.

The first plot shows the merge time (y axis) growth as concurrency (x axis) increases, animated to show the differences for a given number of input sequences (analagous to cluster size).

![thread-groups-typical-distribution-1-8-small](https://user-images.githubusercontent.com/1577461/68105759-6125e880-fe94-11e9-86a4-cae8fb52b92b.gif)

Note that the x axis is the _total_ concurrency count, not the number of threads of this particular group. Also worth pointing out is that the degradation of performance happens at a significantly higher level of concurrency than the previous (unrealistic) worse case performance, but in terms of characteristics, it does share some aspects with the previous plots, such as 8 input sequences being a lot more performant than say 64, and after a certain threshold, the performance of the parallel approach crosses the limit of the same threaded serial merge approach.


The larger 'queries' tell a similar tale:

![thread-groups-typical-distribution-1-8-moderately-large](https://user-images.githubusercontent.com/1577461/68106055-4142f480-fe95-11e9-897b-57c7cf8b4ace.gif)

The differences here when the parallel merge sequence crosses the threshold look to me a fair bit less dramatic than the 'small' sequences, but keep in mind the 'big jump' in the small sequences only amount to a few hundred milliseconds, so it's not quite as dramatic as it appears.

The final plot shows the overall average between both groups:

![thread-groups-typical-distribution-1-8-average](https://user-images.githubusercontent.com/1577461/68105727-46ec0a80-fe94-11e9-9854-aaae9d8405c7.gif)

which I find a bit less useful than the other 2 plots, but included anyway for completeness.

clintropolis commented:
### simulated heavy load
I collected another round of data using the same benchmarks as my 'more realistic worst case' comment, but this time plotting what happens when a large number of queries all start within a 500ms spread, which might be a more typical heavy load, rather than simulating a large concurrent spike of simultaneous queries like the last set of results.

In this scenario, parallel merges outperform the same threaded merges until much higher concurrency than the concurrent spike model.

##### 'small' sequences
![thread-groups-typical-distribution-small-500ms](https://user-images.githubusercontent.com/1577461/68194001-09a57c80-ff69-11e9-9325-d3e70b4853d2.gif)

##### 'moderately large' sequences
![thread-groups-typical-distribution-moderately-large-500ms](https://user-images.githubusercontent.com/1577461/68193956-f7c3d980-ff68-11e9-9d29-d3ee688d8cb6.gif)


##### overall average
![thread-groups-typical-distribution-average-500ms](https://user-images.githubusercontent.com/1577461/68194026-132ee480-ff69-11e9-9601-921a98188feb.gif)

I think future work could focus on making the concurrent spike behavior a bit more chillax through a variety of means, but I find these results to be 'good enough' for now. 

Anyone want to see any other scenarios?

himanshug commented:
@clintropolis thanks for all the benchmarks, I haven't had the opportunity to look at the new developments yet but get back to reviewing this week.

one thing, I am not sure if taken care or not, many people run the druid processes inside docker containers where Runtime.getRuntime().availableProcessors() returns the available processes from host and not from the "container". ( https://bugs.openjdk.java.net/browse/JDK-8140793 , I think it has been changed in jdk10) .  Given the sensitivity of performance to availableProcessors() returned value, it might be good to make that area a bit configurable if not already. I will hopefully offer more specific suggestion when reviewing again.

clintropolis commented:
>@clintropolis thanks for all the benchmarks, I haven't had the opportunity to look at the new developments yet but get back to reviewing this week.

Yeah, no problem, thanks for asking the hard questions to make me collect them, the result is the PR is in a better state than before them :metal:. The production part of the code hasn't really changed much in the last couple of weeks other than a few lines to change behavior of the parallelism computing method, and mostly changes to the default values.

>Given the sensitivity of performance to availableProcessors() returned value, it might be good to make that area a bit configurable if not already. I will hopefully offer more specific suggestion when reviewing again.

This stuff should all be controllable via configs, `druid.processing.merge.pool.parallelism` to control the FJP pool size, and `druid.merge.pool.defaultMaxQueryParallelism` to control individual query max parallelism (this one can also be set on the query context through `parallelMergeParallelism`).

clintropolis commented:
Thanks for the review everyone!
